Adding vector functions (#134)
tigergraph/gsql-docs: modules/querying/pages/func/vector-functions.adoc

= Vector Functions
:experimental:

Vector Functions in TigerGraph are currently implemented as UDFs and preloaded as part of the Graph Data Science Library:

> xref:3.9@graph-ml:similarity-algorithms:index.adoc[]

Reviewer: This is a 3.10.1 PR, but your link path points to 3.9.

Author: Thank you. This has been updated.

> == Vector

Reviewer: This page is in the GSQL **Functions** documentation. It is not for **algorithms**, so do not copy the descriptions of similarity algorithms here.

My original suggestion was not to document the similarity functions here, but to direct the reader to the GDS Similarity Algorithms page, and to put the documentation there.

Author: I understand now. I will leave these here though. I don't think it is a bad idea to have just the functions listed here.

=== tg_euclidean_distance_accum ()
====  Syntax

> `tg_euclidean_distance_accum (VectorA, VectorB)`

Reviewer: Note that it is now `tg_similarity_accum (VectorA, VectorB, "EUCLIDEAN")`

Author: Thank you this has been updated with latest commit.

====  Description
The Euclidean distance between two vectors.

====  Return type
`double`

====  Parameters
|===
| Parameter | Description | Data type

| VectorA
| An n-dimensional vector denoted by a `+ListAccum+` of length `n`
| `+ListAccum<INT/UINT/FLOAT/DOUBLE>+`

| VectorB
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`
|`ListAccum<INT/UINT/FLOAT/DOUBLE>`
|===

====  Example
[tabs]
====
Query::
+
--
[,gsql]
----
CREATE QUERY euclidean_example() FOR GRAPH social {
    ListAccum<INT> @@a = [1, 2, 3];
    ListAccum<INT> @@b = [4, 5, 6];

>     double distance = tg_euclidean_distance_accum(@@a, @@b);

Reviewer: ```suggestion
    double distance = tg_similarity_accum(@@a, @@b, "EUCLIDEAN");
```

Author: This has been updated with latest commit.

    PRINT distance;
}
----
--
Result::
+
--
[,json]
----
{
    "distance": 5.19615
}
----
--
====

> == Vector Similarity

Reviewer: This seems to an extra heading. You have 1 of the 5 above, and 4 of the 5 below. There is no reason to separate them with this heading.

Author: Thank you. This was removed.

|===

==== Example
[tabs]
====
Query::
+
--
[,gsql]
----
CREATE QUERY cosine_similarity_example() FOR GRAPH social {
  ListAccum<INT> @@a = [1, 0, 3];
  ListAccum<INT> @@b = [0, 2, 6];
  double similarity = tg_cosine_similarity_accum(@@a, @@b);
  PRINT similarity;
}
----
--
Result::
+
--
[,json]
----
{
    "similarity": 0.868243
}
----
--
====

=== tg_jaccard_similarity()
====  Syntax
`tg_jaccard_similarity()(VectorA, VectorB)`

====  Description

The Jaccard similarity between the two sets.

====  Return type

`double`

====  Parameters
|===
| Parameter | Description | Data type

|`VectorA`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`
|`ListAccum<INT/UINT/FLOAT/DOUBLE>`

|`VectorB`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`
|`ListAccum<INT/UINT/FLOAT/DOUBLE>`
|===

==== Example
[tabs]
====
Query::
+
--
[,gsql]
----
CREATE QUERY jaccard_similarity_example() FOR GRAPH social {
  ListAccum<INT> @@a = [1, 2, 3];
  ListAccum<INT> @@b = [2, 3, 4];
  double similarity = tg_jaccard_similarity_accum(@@a, @@b);
  PRINT similarity;
}
----
--
Result::
+
--
[,json]
----
{
    "jaccard_similarity": 0.5
}
----
--
====

> == Vector Similarity of Neighborhoods

Reviewer: Do not include this section.

Author: This was committed out with latest commit.

> = Vector Similarity Function

Reviewer: ```suggestion
= Vector Functions
```
This page should begin the same way that the other GSQL function pages start. We are presenting a category of functions; the category is usually defined by the type of data that is works with, in this case, vectors.  In the long run, there might be several functions on this page, so don't speak as those it's just one function.

:experimental:

> Vector Similarity Function in TigerGraph is currently implemented as UDFs and preloaded as part of the Graph Data Science Library:

Reviewer: ```suggestion
This page lists all the functions in the GSQL query language that operate on vector data.  In general, a vector could be either a `LIST<>` type attribute or a `ListAccum<>` variable. 
```

xref:graph-ml:similarity-algorithms:index.adoc[]

[CAUTION]
====
Vector Similarity Functions are subcomponents for the Appropriate Nearest Nearest (ANN) algorithm which is still in beta.

> ====

Reviewer: ```suggestion
[IMPORTANT]
====
The current vector functions are implemented as UDFs and preloaded as part of the Graph Data Science Library and thus are not yet permanently parts of the GSQL language.
The current vector functions work with` ListAccum<>` inputs but not directly with `List<>` attributes.
====
```

> [IMPORTANT]
> ====
> Beta features are for personal or R&D use and not for production use.
> ====
> 

Reviewer: ```suggestion
```

== tg_similarity_accum()

> The function takes 3 parameters.

Reviewer: ```suggestion
Returns a similarity (or distance) measure between two vectors.
This function supports multiple similarity measures; the user selects which measure by setting one of the input parameters.
```

> * The first 2 parameters are `ListAccum<>` for the two vectors being compared.
> * The 3rd parameter is a string and indicates which variation of the function to use.

Reviewer: ```suggestion
* The first two parameters are `ListAccum<>` for the two vectors being compared.
* The third parameter is a string and indicates which variation of the function to use.
```

+

> ====
> .For example:
> 
> ** If the string’s value is `“COSINE”`, then the function performs cosine similarity.
> ** If the string’s value is `“JACCARD”`, it performs Jaccard similarity.
> ====

Reviewer: ```suggestion
|===
| Value | Function

| "COSINE"
| cosine similarity

| "EUCLIDEAN"
| Euclidean distance

| "JACCARD"
| Jaccard similarity

| "OVERLAP"
| overlap similarity

| "PEARSON"
| Pearson correlation coefficient
|===
```

=== Euclidean Distance
====  Syntax
`tg_similarity_accum(VectorA, VectorB, "EUCLIDEAN")`

====  Description

> The Euclidean distance between two vectors.

Reviewer: ```suggestion
The Euclidean distance between two vectors.
The vectors must have the same length and have corresponding numeric elements.
```

====  Return type
`double`

====  Parameters
|===
| Parameter | Description | Data type

| VectorA
| An n-dimensional vector denoted by a `+ListAccum+` of length `n`.
| `+ListAccum<INT/UINT/FLOAT/DOUBLE>+`

| VectorB
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`.
|`ListAccum<INT/UINT/FLOAT/DOUBLE>`

| "EUCLIDEAN"
| A string that identifies the similarity function type.
| `const string`

|===

====  Example
[tabs]
====
Query::
+
--
[,gsql]
----
CREATE QUERY euclidean_example() FOR GRAPH social {
    ListAccum<INT> @@a = [1, 2, 3];
    ListAccum<INT> @@b = [4, 5, 6];
    double distance = tg_similarity_accum(@@a, @@b, "EUCLIDEAN");
    PRINT distance;
}
----
--
Result::
+
--
[,json]
----
{
    "distance": 5.19615
}
----
--
====

=== Overlap
==== Syntax
`tg_similarity_accum( VectorA, VectorB, "OVERLAP" )`

====  Description

> The overlap coefficient between two vectors.

Reviewer: ```suggestion
The overlap coefficient between two vectors, that is, the number of items appearing in both lists, divided by the number of items in the shorter list.
Overlap similarity is intended for vectors which are unordered sets of _categorical data_, such as the sets of city names mentioned in travel blogs.
```

==== Return type
`double`

====  Parameters
|===
| Parameter | Description | Data type

|`VectorA`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`.

> |`+ListAccum<INT/UINT/FLOAT/DOUBLE>+`

Reviewer: ```suggestion
|`+ListAccum<INT/UINT/STRING>+`
```

|`VectorB`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`.

> |`+ListAccum<INT/UINT/FLOAT/DOUBLE>+`

Reviewer: ```suggestion
|`+ListAccum<INT/UINT/STRING>+`
```

| "OVERLAP"
| A string that identifies the similarity function type.
| `const string`
|===

==== Example
[tabs]
====
Query::
+
--
[,gsql]
----
CREATE QUERY overlap_example(/* Parameters here */) FOR GRAPH social {
    ListAccum<INT> @@a = [1, 2, 3];
    ListAccum<INT> @@b = [2, 2, 3];
    double overlap_similarity = tg_similarity_accum(@@a, @@b, "OVERLAP");
    PRINT overlap_similarity;
}
----
--
Result::
+
--
[,json]
----
[
  {
    "overlap_similarity": 0.66667
  }
]
----
--
====

=== Pearson
====  Syntax
`tg_similarity_accum(VectorA, VectorB, "PEARSON")`

====  Description

> The Pearson correlation coefficient between two vectors.

Reviewer: ```suggestion
The Pearson correlation coefficient between two vectors.
The vectors must have the same length and have corresponding numeric elements.
```

====  Return type
`double`

====  Parameters
|===
| Parameter | Description | Data type

|`VectorA`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`.
|`ListAccum<INT/UINT/FLOAT/DOUBLE>`

|`VectorB`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`.
|`ListAccum<INT/UINT/FLOAT/DOUBLE>`

| "PEARSON"
| A string that identifies the similarity function type.
| `const string`
|===

==== Example
[tabs]
====
Query::
+
--
[,gsql]
----
CREATE QUERY pearson_example() FOR GRAPH social {
    ListAccum<INT> @@a = [1, 2, 3];
    ListAccum<INT> @@b = [2, 2, 3];
    double pearson_similarity = tg_similarity_accum(@@a, @@b "PEARSON");
    PRINT pearson_similarity;
}
----
--
Result::
+
--
[,json]
----
{
    "pearson_similarity": 0.86603
}
----
--
====

=== Cosine
====  Syntax
`tg_similarity_accum(VectorA, VectorB, "COSINE")`

====  Description

> The cosine similarity between the two vectors.

Reviewer: ```suggestion
The cosine similarity between the two vectors.
The vectors must have the same length and have corresponding numeric elements.
```

====  Return type

`double`

====  Parameters
|===
| Parameter | Description | Data type

|`VectorA`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`.
|`ListAccum<INT/UINT/FLOAT/DOUBLE>`

|`VectorB`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`.
|`ListAccum<INT/UINT/FLOAT/DOUBLE>`

| "COSINE"
| A string that identifies the similarity function type.
| `const string`
|===

==== Example
[tabs]
====
Query::
+
--
[,gsql]
----
CREATE QUERY cosine_similarity_example() FOR GRAPH social {
  ListAccum<INT> @@a = [1, 0, 3];
  ListAccum<INT> @@b = [0, 2, 6];
  double similarity = tg_similarity_accum(@@a, @@b "COSINE");
  PRINT similarity;
}
----
--
Result::
+
--
[,json]
----
{
    "similarity": 0.868243
}
----
--
====

=== Jaccard
====  Syntax
`tg_similarity_accum(VectorA, VectorB "JACCARD")`

====  Description

> The Jaccard similarity between the two sets.

Reviewer: ```suggestion
The Jaccard similarity between the two sets, that is, the number of items appearing in both sets, divided by the number of unique items in the lists.
Jaccard similarity, like overlap similarity, is intended for vectors which are unordered sets of _categorical data_, such as the sets of city names mentioned in travel blogs.
```

====  Return type

`double`

====  Parameters
|===
| Parameter | Description | Data type

|`VectorA`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`.

> |`ListAccum<INT/UINT/FLOAT/DOUBLE>`

Reviewer: ```suggestion
|`ListAccum<INT/UINT/STRING>`
```

|`VectorB`
|An n-dimensional vector denoted by a `+ListAccum+` of length `n`.

> |`ListAccum<INT/UINT/FLOAT/DOUBLE>`

Reviewer: ```suggestion
|`ListAccum<INT/UINT/STRING>`
```
